### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by marcinlechnik**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,10 +19,15 @@
     "url": "https://github.com/strudeljs/strudel-redux/issues"
   },
   "homepage": "https://github.com/strudeljs/strudel-redux#readme",
-  "dependencies": {},
+  "dependencies": {
+    "@babel/preset-env": "^7.3.1",
```
Comment (xolir) [inline package.json]: This should be moved into devDependencies

File: package.json [hunk 2]
```diff
@@ -17,10 +19,15 @@
...
+    "redux": "^4.0.1"
+  },
   "devDependencies": {
     "babel-core": "^6.26.0",
     "babel-preset-es2015-rollup": "^3.0.0",
+    "jest": "^24.1.0",
+    "redux-mock-store": "^1.5.3",
```
Comment (xolir) [inline package.json]: This should be removed altogether

File: src/subscribe.js [hunk 3]
```diff
@@ -1,67 +1,42 @@
...
-const mixin = (target) => {
-  bindedMethods.forEach((funcName) => {
-    patch(target, funcName);
+export const subscribedStateChanged = (observedState, stateMemory) => Object
+  .entries(observedState)
+  .some(([key, value]) => stateMemory[key] !== value);
+
+/* eslint-disable no-param-reassign, func-names */
+export const Subscribe = (store, {observed, statics: () => {} }) => function (target) {
+  debugger;
```
Comment (xolir) [inline src/subscribe.js]: This should be removed

File: src/subscribe.js [hunk 4]
```diff
@@ -1,67 +1,42 @@
...
-  if (!target || !target.isStrudelClass) {
-    throw new Error('Please pass a valid component to "Subscribe"');
-  }
+  const orgInit = target.prototype.init;
+
+  target.prototype.init = function () {
+    orgInit.call(this);
+
+    store.subscribe(() => {
+      debugger;
```
Comment (xolir) [inline src/subscribe.js]: Remove this statement

File: src/subscribe.js [hunk 5]
```diff
@@ -1,67 +1,42 @@
...
+      const storeState = store.getState();
+      const observedState = observed(storeState);
+      const stateChanged = subscribedStateChanged(observedState, stateMemory);
+      
+      debugger;
```
Comment (xolir) [inline src/subscribe.js]: Remove this statement

File: src/subscribe.js [hunk 6]
```diff
@@ -1,67 +1,42 @@
...
+      if (stateChanged) {
+        debugger;
```
Comment (xolir) [inline src/subscribe.js]: Remove this statement

File: src/subscribe.spec.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    });
+  
+    test('should return true if states differ', () => {
+      expect(subscribedStateChanged(initialStoreState, secondaryStoreState)).toEqual(true);
+    });
+
+    test('should return observed and static state', () => {
+      class TestComponent {
+        onStateChange() {
+          debugger;
```
Comment (xolir) [inline src/subscribe.spec.js]: Remove this statement

File: src/subscribe.spec.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      component.init();
+
+      store.dispatch({ type: actions.observed});
+
+      expect(spy).toHaveBeenCalledWith({
+        price: 140,
+        stockInfo: 'InStock',
+      });
+
+      store.dispatch({ type: actions.observed});
```
Comment (kulek1) [inline src/subscribe.spec.js]:
```suggestion
      store.dispatch({ type: actions.observed });
```

File: src/subscribe.spec.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+      expect(spy).toHaveBeenCalledTimes(2);
+
+      store.dispatch({ type: actions.static });
+      expect(spy).toHaveBeenCalledTimes(2);
+
+      store.dispatch({ type: actions.observed});
```
Comment (kulek1) [inline src/subscribe.spec.js]:
```suggestion
      store.dispatch({ type: actions.observed });
```

**Comment by mateuszluczak** [comment]
@xolir @marcinlechnik 
Any plans to merge and release this?

**Pull request merged**